Make it easy to find out the k8s and flux version

opudrovs opened the pull request:
Closes #2285
Closes #2328 

- ~~I've added a namespace to the `GetVersionRequest` object to make `TestGetVersion` pass because it could not find the `flux-system` namespace without it. So, for the test I create a new namespace or if the namespace is not provided default to `flux-system` (`wego.DefaultNamespace` constant). If anyone knows how to improve this and the issue with nested ifs (will add a comment), I'm all ears. 🙂~~

UPDATE: Removed the namespace field from `GetVersionRequest` (switched to listing namespaces) and fixed nested ifs based on @ozamosi 's suggestions.

- I've copy-pasted the test for `GetImpersonatedDiscoveryClient` from the test for `GetImpersonatedClient`. I am not sure what this API does and what exactly should be tested, maybe it should be modified.

Also removed the following from the copied test, maybe it should be put back (this part was failing because the number of found namespaces in the copy-pasted part was 0 instead of 1 or 2):

```
t.Run("checks all namespaces in the cluster when through the filtering", func(t *testing.T) {
	g.Expect(nsChecker.FilterAccessibleNamespacesCallCount()).To(Equal(1))

	_, _, nss := nsChecker.FilterAccessibleNamespacesArgsForCall(0)
	nsFound := 0
	for _, n := range nss {
		if n.Name == ns1.Name || n.Name == ns2.Name {
			nsFound++
		}
	}

	g.Expect(nsFound).To(Equal(2))
})
```

opudrovs commented on core/server/version.go:
```diff
@@ -15,10 +21,66 @@ var (
+				fluxVersion = u.GetLabels()["app.kubernetes.io/version"]
+			}
+		}
+	}
```
@ozamosi @chanwit suggestions on how to simplify nested ifs here (looks similar to a "callback hell" in JS) welcome!

I don't want to return nil from `GetVersion` on error when getting additional values, because I want to display at least available values (at least the values set at the build time) in the app for debug purposes.

ozamosi replied:
Maybe lift this bit into a new function, return an error from there any time there's an error, and then this function only has one error to check?

opudrovs replied:
Thanks, good idea, will do it.

opudrovs commented:
@ozamosi @chanwit also, maybe I can use the `VersionLabelKey` from the flux.go instead of `app.kubernetes.io/version` to get the label value?

ozamosi commented on core/server/version.go:
```diff
@@ -15,10 +21,66 @@ var (
+
+	key := client.ObjectKey{
+		Name: ns,
+	}
```
So, you're right that this can be pretty much any namespace - but I'm not sure the frontend knows what namespace to use :thinking: 

What if instead of `Get`ing a specific namespace, we `List` filtered by `{"app.kuberenetes.io/part-of": "flux"}` and just grab the first result we find? After a bit of grepping - I think this helper does the filtering? https://github.com/weaveworks/weave-gitops/blob/main/core/server/server.go#L86

opudrovs replied:
OK, can do it.

From frontend we don't provide any namespace. This param was added just for passing the test (because when running the test it could not find the "flux-system" namespace). If the namespace is not provided (like when we make the request via the API, `{}` with no args), we just default to "flux-system" (and while the app is running normally, the "flux-system" namespace is found).

I need to see if the list helper will return the flux-system/flux-whatever namespace while running the test.

ozamosi commented:
> @ozamosi @chanwit also, maybe I can use the `VersionLabelKey` from the flux.go instead of `app.kubernetes.io/version` to get the label value?

Sounds good!

chanwit commented on core/server/version.go:
```diff
@@ -14,11 +22,81 @@ var (
+	if err != nil {
+		return defaultVersion, fmt.Errorf("error getting object: %w", err)
+	} else {
+		return u.GetLabels()["app.kubernetes.io/version"], nil
```
u.GetLabels() may return `nil` here.

opudrovs replied:
Thanks, will keep it in mind.

opudrovs commented:
@ozamosi @chanwit added grabbing a namespace from the list of namespaces. For testing, I create a new namespace with required key because the `flux-ns` namespace available during testing doesn't have a Flux version.

Also moved creating scoped client, getting Flux and K8s version to separate functions, based on @ozamosi 's suggestion.

@chanwit just in case added checking for nil when getting the version from labels.

@joshri I've added displaying Flux and Kubernetes versions in the footer. Also added `noWrap` to the Text component and  removed neutral30 color specified for some footer elements because it is already specified for the whole footer, it looks like, the elements already inherit this color from footer. If it is wrong, can move it back.

Spaced elements based on the draft design in the UI ticket description to make it consistent with current styling of the Footer component. Please let me know if styling/spacing/whatever is wrong.

opudrovs commented:
@ozamosi @chanwit also, would it be possible to filter for Version label key with any value? Or only filtering by specific values is possible?

```
opts := client.MatchingLabels{
    coretypes.PartOfLabel: FluxNamespacePartOf,
    flux.VersionLabelKey: <any value>
}
```

Right now I am filtering labels only by `coretypes.PartOfLabel: FluxNamespacePartOf,` and then loop through the result and find a label which has `flux.VersionLabelKey` key.

opudrovs commented:
@ozamosi OK, thank you, got it, will add the changes now.

joshri commented on ui/components/Footer.tsx:
```diff
@@ -37,23 +37,54 @@ function Footer({ className }: Props) {
+    ) : null;
+
+  const kubeVersionText = versionData.kubeVersion;
+  const kubeVersion =
```
fluxVersion and kubeVersion could potentially be combined into one component with props but I don't mind either way

opudrovs replied:
Added the Version component with snapshot tests.

1) Not sure if adding `attrs` when exported is still required, like recommended in the style.md doc, so still added it like that:

```
export default styled(Version).attrs({ className: Version.name })``;
```

2) Footer versions are wrapped at narrow screens (see layout at 768 or 1000 px for example) , otherwise we would have to add a horiz scrollbar, preferably for the whole layout (or it looks ugly if added only to the footer).

Does this look OK?

<img width="735" alt="Screenshot 2022-06-28 at 08 38 53" src="https://user-images.githubusercontent.com/8824708/176110484-19bad6fe-e43e-49cb-a810-42027bc45749.png">

joshri replied:
Three options that come to mind:
1. Set everything to nowrap and just have it overflow the container 
2. Increase the minWidth of the container so we know we're always fitting the whole footer in
3. Decrease font size on the footer?
Maybe we do a mix of all three?

opudrovs replied:
Used options 1 and 2: set the Version component text to nowrap and changed 768px in Layout.tsx from 768px to 900px (conveniently, a Material-UI breakpoints).

Everything I tried with scrollbars in the footer and footer containers looked ugly. So, I've decided not to add `overflow-x: auto`.

I think it's either fixes or we'll need to serious redesign the footer (for example, move it under the whole app content or move content info + email text somewhere from the footer), because its current text is really wide.

So, the current version looks best for now.

joshri commented on ui/components/Footer.tsx:
```diff
@@ -37,23 +37,54 @@ function Footer({ className }: Props) {
     <Flex as="footer" wide between className={className} role="footer">
       <LeftFoot>
-        <Text color="neutral30">Need help? Contact us at</Text>
+        <Text>Need help? Contact us at</Text>
```
add the excellent `noWrap` prop here

opudrovs replied:
Added.

joshri commented on ui/components/Footer.tsx:
```diff
@@ -37,23 +37,54 @@ function Footer({ className }: Props) {
         )}
         <Spacer padding="xxs" />
-        <Text color="neutral30">© 2022 Weaveworks</Text>
+        <Text>© 2022 Weaveworks</Text>
```
same as above - add the excellent `noWrap` prop!!

joshri reviewed (approved):
I had to update something...classic. Rebuilt and the footer looks good!
